edb/server/protocol/auth_ext/__init__.py

```diff
@@ -0,0 +1,86 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2022-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import http
+import json
+import urllib.parse
+
+from edb import errors
+from edb.common import debug
+from edb.common import markup
+
+from . import oauth
+
+
+async def handle_request(
+    request, response, db, extension_base_path: str, args: list[str]
+):
+    try:
+        # Set up routing to the appropriate handler
+        if args[0] == "authorize":
+            provider_name: str | None = _get_search_param(
+                request.url.query.decode("ascii"), "provider"
```

Comment:
Question: why we chose API like `/db/{dbname}/ext/auth/authorize?provider=github` instead of `/db/{dbname}/ext/auth/github/authorize`? The same for the callback URL. The former is assuming we will always have `/authorize` and `/callback` for any providers.

Reply:
Adding the provider to the search params feels lighter-weight and the callback itself doesn't need to refer to the provider since we don't want to get the provider name from the URL but rather from the `provider` claim within the state JWT. Doing it this way means that we avoid the case where we accidentally copy-pasta-fail the callback URL structure from GitHub to GitLab and don't notice that the URL is out of sync with the state variable. In that case we have to decide whether that's a fatal error, or just an accident we don't really care about.

Reply:
I see. I don't like long URLs either.

But I'm still a bit concerned about the API tbh. What if a future provider has a different `/callback` args format that we can't simply reuse the same handler method? If we will have `/callback/provider8`, I'd rather make it `/provider8/callback` now and tolerate the length of the URL. 🤔

Also I'm not sure why we have to sign the `provider` string into the JWT. If the attacker manipulated the provider, the final access token exchange will fail anyways, won't it? Also the provider is something configured inside the system, the attacker won't be able to redirect to a non-existing one.

Anyways, I think it's okay now that we only have OAuth 2.0 and stay with shorter `/authorize` and `/callback`, and let the future providers worry about their own URLs.

Reply:
> What if a future provider has a different /callback args format that we can't simply reuse the same handler method?

We specify the callback URL in the `redirect_uri`, so we control this. And each provider has some slightly different (but very similar) set of query parameters that they will add to this URL or in some cases that data will be provided as form data in a POST request. Each `Provider` will know where to look for these values. But as far as these two URLs are concerned we control them entirely: client libraries will know to call the `/authorize?provider` URL in order to start the flow, and the server here specifies the callback URL.

> Also I'm not sure why we have to sign the provider string into the JWT. If the attacker manipulated the provider, the final access token exchange will fail anyways, won't it?

Indeed! But why have to trust the URL when the server itself can authoritatively make the claim that "this state string is intended to be used by this certain provider" and our callback can trust it due to the signature.

> Also the provider is something configured inside the system, the attacker won't be able to redirect to a non-existing one.

Sure this is less of a security issue (as far as I can tell) and more of a correctness issue: if we hardcode the provider into the callback URL itself, we're open to bugs where we have typos or copy-paste mistakes. If we get the provider from the state string, we at least know the value provided there is the same one that we built during the authorize endpoint.

I could see maybe:
- `/db/{db_name}/ext/auth/{provider_name}/authorize`
- `/db/{db_name}/ext/auth/callback`

But that feels surprising to me.

Reply:
OK cool cool!

> And each provider has some slightly different (but very similar) set of query parameters

This is true in the scope of OAuth 2.0 / OpenID Connect. I'm okay with assuming by default `/ext/auth/authorize` and `/ext/auth/callback` are OAuth 2.0 client URLs.

> Each `Provider` will know where to look for these values.

We'll need "these values" first to know which `Provider` to use in case of `/callback` ;)

---

Oh actually that reminds me of another thing - will EdgeDB auth ext itself become a provider? I have impression we mentioned the Implicit Flow or PKCE stuff for user application login in the meetings, but I'm not sure if we decided on anything yet. I'm just thinking if there could be any URL/prefix conflict. But it shouldn't be a big problem.

Reply:
> We'll need "these values" first to know which Provider to use in case of /callback ;)

Not as it's currently written: we get the `Provider` from the JWT claims in the `state` query parameter (or form data if it's a POST), and then evaluate the rest of the request with the `Provider`. True that it's specific to OAuth 2.0 but I think the current scope (no pun intended) is to get a single provider working (GitHub in this case) and add more as we go. If we need to support OAuth 1.0a for some reason (doesn't seem like there are any providers we should spend time on that are 1.0a only) I think it should be handled by a separate set of routes?

Maybe there are other auth schemes I haven't really explored much here (SAML maybe?) that makes this less compatible?

> Oh actually that reminds me of another thing - will EdgeDB auth ext itself become a provider? I have impression we mentioned the Implicit Flow or PKCE stuff for user application login in the meetings, but I'm not sure if we decided on anything yet. I'm just thinking if there could be any URL/prefix conflict. But it shouldn't be a big problem.

Yeah, I absolutely want to support username+password as a kind of self-hosted auth server that maybe skips the OAuth flow entirely but writes a similar `Profile` object to the database. I haven't really set about tackling this, but I think it should be the second provider we try to configure. FWIW, I don't think we need to consider this "provider" as a full-on OIDC style identity provider for any external services, so I suspect the actual implementation will just be `authorize?provider=email` (or maybe `password`) and then have the server just short-circuit the whole back-and-forth with itself and provide a signed OIDC identity on sucessful login. Sign up will probably be the big lift there, but I haven't thought that through yet.

---

Luckily, we have time to go in a few different directions to see if we end up with conflicts and I feel pretty free to change everything here at-will until we ship.